### ISSUE TITLE
unexpected synthetic: `SyntheticSyntax`

### ISSUE THREAD

**Issue by griffi-gh** [rust-lang/rust-analyzer#16620]
```
Panic context:
> fetch_native_diagnostics

thread 'Worker' panicked at crates\hir\src\diagnostics.rs:521:63:
unexpected synthetic: SyntheticSyntax
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:645
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:72
   2: core::result::unwrap_failed
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\result.rs:1649
   3: hir::diagnostics::AnyDiagnostic::inference_diagnostic
   4: hir::DefWithBody::diagnostics
   5: hir::ModuleDef::diagnostics
   6: hir::Module::diagnostics
   7: ide_diagnostics::diagnostics
   8: salsa::Cancelled::catch
   9: ide::Analysis::diagnostics
  10: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
  11: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
  12: rust_analyzer::diagnostics::fetch_native_diagnostics
  13: core::ops::function::FnOnce::call_once{{vtable.shim}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Panic context:
> fetch_native_diagnostics

thread 'Worker' panicked at crates\hir\src\diagnostics.rs:521:63:
unexpected synthetic: SyntheticSyntax
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:645
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:72
   2: core::result::unwrap_failed
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\result.rs:1649
   3: hir::diagnostics::AnyDiagnostic::inference_diagnostic
   4: hir::DefWithBody::diagnostics
   5: hir::ModuleDef::diagnostics
   6: hir::Module::diagnostics
   7: ide_diagnostics::diagnostics
   8: salsa::Cancelled::catch
   9: ide::Analysis::diagnostics
  10: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
  11: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
  12: rust_analyzer::diagnostics::fetch_native_diagnostics
  13: core::ops::function::FnOnce::call_once{{vtable.shim}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Panic context:
> fetch_native_diagnostics

thread 'Worker' panicked at crates\hir\src\diagnostics.rs:521:63:
unexpected synthetic: SyntheticSyntax
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:645
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:72
   2: core::result::unwrap_failed
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\result.rs:1649
   3: hir::diagnostics::AnyDiagnostic::inference_diagnostic
   4: hir::DefWithBody::diagnostics
   5: hir::ModuleDef::diagnostics
   6: hir::Module::diagnostics
   7: ide_diagnostics::diagnostics
   8: salsa::Cancelled::catch
   9: ide::Analysis::diagnostics
  10: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
  11: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
  12: rust_analyzer::diagnostics::fetch_native_diagnostics
  13: core::ops::function::FnOnce::call_once{{vtable.shim}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Panic context:
> fetch_native_diagnostics

thread 'Worker' panicked at crates\hir\src\diagnostics.rs:521:63:
unexpected synthetic: SyntheticSyntax
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:645
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:72
   2: core::result::unwrap_failed
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\result.rs:1649
   3: hir::diagnostics::AnyDiagnostic::inference_diagnostic
   4: hir::DefWithBody::diagnostics
   5: hir::ModuleDef::diagnostics
   6: hir::Module::diagnostics
   7: ide_diagnostics::diagnostics
   8: salsa::Cancelled::catch
   9: ide::Analysis::diagnostics
  10: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
  11: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
  12: rust_analyzer::diagnostics::fetch_native_diagnostics
  13: core::ops::function::FnOnce::call_once{{vtable.shim}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**rust-analyzer version**: rust-analyzer version: 0.4.1852-standalone (9f0495761 2024-02-20)

**rustc version**: rustc 1.78.0-nightly (3246e7951 2024-02-19)

**relevant settings**: -

**Comment by Veykril** [comment]
Needs a snippet to reproduce

**Comment by poliorcetics** [comment]
I got it on https://github.com/starship/starship, just opening the project

My RA-config is:

```
assist.importGranularity = "module"
cargo.extraEnv.CARGO_TARGET_DIR = "/Users/alexis/.local/cache/rust-analyzer-target-dir"
check.command = "clippy"
completion.fullFunctionSignatures.enable = true
diagnostics.disabled = ["inactive-code", "inactive_code", "unresolved-proc-macro", "unresolved_proc_macro"]
hover.actions.references.enable = true
inlayHints.closingBraceHints.minLines = 10
inlayHints.closureReturnTypeHints.enable = "with_block"
inlayHints.discriminantHints.enable = "fieldless"
inlayHints.expressionAdjustmentHints.enable = "never"
inlayHints.expressionAdjustmentHints.hideOutsideUnsafe = false
inlayHints.expressionAdjustmentHints.mode = "prefer_prefix"
inlayHints.lifetimeElisionHints.enable = "skip_trivial"
inlayHints.typeHints.hideClosureInitialization = false
lens.references.adt.enable = true
lens.references.enumVariant.enable = true
lens.references.method.enable = true
lens.references.trait.enable = true
lruCapacity = 256
workspace.symbol.search.kind = "all_symbols"
workspace.symbol.search.limit = 128
workspace.symbol.search.scope = "workspace"
```

**Comment by poliorcetics** [comment]
I'm on macOS, with commit 543d7e98d

**Comment by Veykril** [comment]
Fixed in #16684
